src/core/event_engine/effect_dispatcher.rs

```diff
@@ -0,0 +1,94 @@
+use crate::{
+    core::event_engine::{Effect, EffectHandler, EffectInvocation, Event},
+    lib::alloc::{rc::Rc, vec, vec::Vec},
+};
+use phantom_type::PhantomType;
+use spin::rwlock::RwLock;
+
+/// State machine effects dispatcher.
+#[allow(dead_code)]
+pub(crate) struct EffectDispatcher<EH, EF, EI>
+where
+    EI: EffectInvocation<Effect = EF>,
+    EH: EffectHandler<EI, EF>,
+    EF: Effect,
+{
+    /// Effect invocation handler.
+    ///
+    /// Handler responsible for providing actual implementation of
+    handler: EH,
+
+    /// Dispatched effects managed by dispatcher.
+    ///
+    /// There are effects whose lifetime should be managed by the dispatcher.
+    /// State machines may have some effects that are exclusive and can only run
+    /// one type of them at once. The dispatcher handles such effects
+    /// and cancels them when required.
+    managed: RwLock<Vec<Rc<EF>>>,
+
+    _invocation: PhantomType<EI>,
+}
+
+impl<EH, EF, EI> EffectDispatcher<EH, EF, EI>
+where
+    EI: EffectInvocation<Effect = EF>,
+    EH: EffectHandler<EI, EF>,
+    EF: Effect,
+{
+    /// Create new effects dispatcher.
+    pub fn new(handler: EH) -> Self {
+        EffectDispatcher {
+            handler,
+            managed: RwLock::new(vec![]),
+            _invocation: Default::default(),
+        }
+    }
+
+    /// Dispatch effect associated with `invocation`.
+    pub fn dispatch<F, E>(&self, invocation: &EI, f: F)
+    where
+        E: Event,
+        F: Fn(Vec<E>),
+    {
+        if let Some(effect) = self.handler.create(invocation) {
+            let effect = Rc::new(effect);
```

Comment:
> Why do we need that RC here?

I have a requirement to store managed `effect` in a vector and make it possible to remove the same effect from list after `run` completion.
I tried to filter in completion block, but borrow checker won't let me go this way. Furthermore, I used `Rc` just to make clone once to add it vector and don't need `Rc` functionality after that.

> Why do we passing the remove_managed_effect into the effect and call it inside of it? 

I can't follow this one.
The idea was that `run` function of effect has completion closure, with which we will be able to remove effect from the list of managed effects. 

> Additionally, if we need reference counter it probably won't work with our client as it needs to be thread safe.
> Use Arc instead.

`Rc` functionality used once (I mean clone) within `dispatch` function, and that is all. All other access to it through vector which stores it should be safe enough because of `RwLock`.

Reply:
>The idea was that run function of effect has completion closure, with which we will be able to remove effect from the list of managed effects.
This one is still confusing to me. 
How effects are allowed to managed the call?
Currently in SubscribeEffect it doesn't make any decision about it. Do you think that some effects will affect that?
We still can casually call that line after calling `run` function.